crates/matrix-sdk/src/room_directory_search.rs

```diff
@@ -0,0 +1,117 @@
+// limitations under the License.
+
+use eyeball_im::{ObservableVector, VectorDiff, VectorSubscriber};
+use futures_core::Stream;
+use ruma::{
+    api::client::directory::get_public_rooms_filtered::v3::Request as PublicRoomsFilterRequest,
+    directory::{Filter, PublicRoomJoinRule},
+    OwnedMxcUri, OwnedRoomAliasId, OwnedRoomId,
+};
+
+use crate::{Client, Result};
+
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct RoomDescription {
+    pub room_id: OwnedRoomId,
+    pub name: Option<String>,
+    pub topic: Option<String>,
+    pub alias: Option<OwnedRoomAliasId>,
+    pub avatar_url: Option<OwnedMxcUri>,
+    pub join_rule: PublicRoomJoinRule,
+    pub is_world_readable: bool,
+    pub joined_members: u64,
+}
+
+pub struct RoomDirectorySearch {
+    batch_size: u32,
+    filter: Option<String>,
+    next_token: Option<String>,
+    client: Client,
+    results: ObservableVector<RoomDescription>,
+    is_at_last_page: bool,
+}
+
+impl RoomDirectorySearch {
+    pub fn new(client: Client) -> Self {
+        Self {
+            batch_size: 0,
+            filter: None,
+            next_token: None,
+            client,
+            results: ObservableVector::new(),
+            is_at_last_page: false,
+        }
+    }
+
+    pub async fn search(&mut self, filter: Option<String>, batch_size: u32) -> Result<()> {
+        self.filter = filter;
+        self.batch_size = batch_size;
+        self.next_token = None;
+        self.results.clear();
+        self.is_at_last_page = false;
+        self.next_page().await
+    }
+
+    pub async fn next_page(&mut self) -> Result<()> {
+        if self.is_at_last_page {
+            return Ok(());
+        }
+        let mut filter = Filter::new();
+        filter.generic_search_term = self.filter.clone();
+
+        let mut request = PublicRoomsFilterRequest::new();
+        request.filter = filter;
+        request.limit = Some(self.batch_size.into());
+        request.since = self.next_token.clone();
+        let response = self.client.public_rooms_filtered(request).await?;
+        self.next_token = response.next_batch;
+        if self.next_token.is_none() {
+            self.is_at_last_page = true;
+        }
+        self.results.append(
+            response
+                .chunk
+                .into_iter()
+                .map(|room| RoomDescription {
+                    room_id: room.room_id,
+                    name: room.name,
+                    topic: room.topic,
+                    alias: room.canonical_alias,
+                    avatar_url: room.avatar_url,
+                    join_rule: room.join_rule,
+                    is_world_readable: room.world_readable,
+                    joined_members: room.num_joined_members.into(),
+                })
+                .collect(),
+        );
+        Ok(())
+    }
+
+    pub fn results(&self) -> impl Stream<Item = VectorDiff<RoomDescription>> {
+        self.results.subscribe().into_stream()
+    }
```

Comment:
Every time you subscribe to an observable value, you need to return its initial value, along with its stream. Why? Imagine you call `next_page` _and then_ `results`. `results` will return the _updates_, but you will miss the first results from the call of `next_page`.

You should replace `into_stream` by [`into_values_and_stream`](https://docs.rs/eyeball-im/latest/eyeball_im/struct.VectorSubscriber.html#method.into_values_and_stream).

Even better (!), returning a `Stream<Item = VectorDiff<…>>` is nice and all, but when it aims to be used by a UI via FFI (which is likely to be the case here), it may be better to batch the `VectorDiff` together. The idea is to avoid crossing the FFI boundary too many times which has a non-negligible impact the performance. Hopefully for us, we have [`into_values_and_batched_stream`](https://docs.rs/eyeball-im/latest/eyeball_im/struct.VectorSubscriber.html#method.into_values_and_batched_stream). Thus, the code becomes:

```suggestion
    pub fn results(&self) -> (Vector<RoomDescription>, impl Stream<Item = Vec<VectorDiff<RoomDescription>>>) {
        self.results.subscribe().into_values_and_batched_stream()
    }
```

Notice that `Stream`'s item is now `Vec<VectorDiff<…>>`.